tools/update_extern.py

```diff
@@ -0,0 +1,166 @@
+    """
+    with ZipFile(filename, "r") as zip_file:
+        zip_file.extractall("extern_pkg")
+    # Remove the zip file
+    os.remove(filename)
+
+
+def move_files(src: Path, dst: Path):
+    """
+    Move the files from the src to the dst.
+    """
+    # if dst already exists, remove it
+    if os.path.exists(dst):
+        print(f"Removing {dst}")
+    # move the files
+    shutil.move(src, dst)
+    print(f"Moving {src} to {dst}")
+
+
+def update_extern(destination):
+    """
+    Update the files in the sunpy/extern directory.
+    """
+    # If the destination contains the appdirs package, extract the appdirs.py file
+    if destination.split("-")[0] == "appdirs":
+        if os.path.exists(f"extern_pkg/{destination}/appdirs.py"):
+            # Move the file to the sunpy/extern directory
+            move_files(f"extern_pkg/{destination}/appdirs.py", SUNPY_DIR / "sunpy" / "extern" / "appdirs.py")
+
+    # If the destination contains the distro package, extract the distro.py file
+    if destination.split("-")[0] == "distro":
+        # Print the path of distro.py
+        if os.path.exists(f"extern_pkg/{destination}/src/distro/distro.py"):
+            # Move the file to the sunpy/extern directory
+            move_files(f"extern_pkg/{destination}/src/distro/distro.py",
+                       SUNPY_DIR / "sunpy" / "extern" / "distro.py")
+
+    # If the destination contains the inflect package, extract the inflect.py file
+    if destination.split("-")[0] == "inflect":
+        if os.path.exists(f"extern_pkg/{destination}/inflect/__init__.py"):
+            # Rename the file to inflect.py
+            os.rename(f"extern_pkg/{destination}/inflect/__init__.py",
+                      f"extern_pkg/{destination}/inflect/inflect.py")
+            # Move the file to the sunpy/extern directory
+            move_files(f"extern_pkg/{destination}/inflect/inflect.py",
+                       SUNPY_DIR / "sunpy" / "extern" / "inflect.py")
+
+    # If the destination contains the parse package, extract the parse.py file
+    if destination.split("-")[0] == "parse":
+        if os.path.exists(f"extern_pkg/{destination}/parse.py"):
+            # Move the file to the sunpy/extern directory
+            move_files(f"extern_pkg/{destination}/parse.py", SUNPY_DIR / "sunpy" / "extern" / "parse.py")
+
+    # Remove the directory
+    shutil.rmtree(f"extern_pkg/{destination}")
+
+
+if __name__ == "__main__":
+
+    # Check if the packages exist
+    if not package_exists():
+        print("Exiting...")
+        exit()
+
+    for package in PACKAGES:
+        # get the url of the package
+        url = get_download_url(PACKAGES[package], package, get_latest_version(PACKAGES[package], package))
+
+        # download the package
+        download_package(url, package)
+
+        # Open the zip file
+        unzip(f"extern_pkg/{package}.zip")
+
+    folders = list()
+    for root, dirs, files in os.walk("extern_pkg"):
+        folders = dirs
+        break
```

Comment:
What does this do?

Reply:
Okay, I will explain what I tried to do.

1. we check that this [dict](https://github.com/sunpy/sunpy/pull/6127/files#diff-bb5c4cdf80ea20b4ae8f8db4b6d8d12e88e1d32f16f2b545ca551accf32f70d4R18) has correct keys and value. If the key-value pair are wrong, show error to the user and exit. 
 ```py
 # Check if the packages exist
    if not package_exists():
        print("Exiting...")
        exit()
```

2. we make a download URL using `get_download_url()` and use that URL to download `.zip` with `download_package()` then unzip it in a temporary directory 
```py
    for package in PACKAGES:
        # get the url of the package
        url = get_download_url(PACKAGES[package], package, get_latest_version(PACKAGES[package], package))

        # download the package
        download_package(url, package)

        # Open the zip file
        unzip(f"extern_pkg/{package}.zip")
```

3. we get names of all the `top-folder` inside that temporary directory in a `list`.
for example,
```
Temporary Directory
├──  appdirs-1.4.4
│  ├──  appdirs.py
│   ...
├──  distro-1.7.0
│  ├──  CHANGELOG.md
│   ...
├──  inflect-5.6.0
│  ├──  CHANGES.rst
│   ...
└──  parse-1.19.0
   ├──  LICENSE
   ├──  parse.py
    ...
```
In this dir, we store to names of packages in a list. 
So, the list `folders` will be 
`folders = ["appdirs-1.4.4","distro-1.7.0","inflect-5.6.0","parse-1.19.0"]`

```py
    folders = list()
    for root, dirs, files in os.walk("extern_pkg"):
        folders = dirs
        break
```
And we don't need to sort this. (This was something related to old code where I used PACKAGES and AUTHORS, but now it's a single dictionary, so I will remove this in the next commit.)
```py
    # Sort the packages in alphabetical order
    folders.sort()
```

Now we use that `list` to extract (move) the required python files, that will be done with `update_extern()` function.

```py
    # Extract the files
    for folder in folders:
        update_extern(folder)
```
4. Finally remove the temporary directory.
```py
    # Remove the temporary directory
    shutil.rmtree("extern_pkg")
```

Reply:
For `1.` it might be less code to verify the HTTP response code of the request that's already in `get_latest_version` instead, although as @nabobalis pointed out we should merge some of those functions.

For `3.` it may simplify things if we extracted directly into the `sunpy/extern` directory. For example, using [`extract`](https://docs.python.org/3/library/zipfile.html#zipfile.ZipFile.extract) instead of `extractall` you should be able to extract the single file only with `_______/sunpy/extern` as the path. Although you may need to delete the file if it exists, e.g.:

```py
def update_package(package_name, author, src):
    zip_filename = download_package(author, package_name)
    dest = SUNPY_EXTERN_DIR / f"{package_name}.py"
    if os.path.exists(dest):
        os.remove(dest)
    with ZipFile(zip_filename, "r") as zip_file:
        zip_file.extract(src, path=SUNPY_EXTERN_DIR)


for package_name, (author, src) in PACKAGES.items():
    update_package(package_name, author, src)
```

Reply:
Yeah this is what we should do, to reduce the amount of code and the complexity of it.

Reply:
Sorry for the late reply, I was busy with college assignments. 

If we try to unzip particular files that would be great, and I wrote it!

But what if the `.py` file we unzipped is named `__init__.py`? 
For example, In the case of `inflect` that would cause to replace the different file in `sunpy/extern`
So, can we extract it in the same temporary directory, rename and then move them?

Reply:
I doubt if it is possible to rename files during their extraction.

Reply:
Yeah, good point! That should work, if you extract without the path you should then be able to move the file separately to the extern directory. Updated example from above:
```py
def update_package(package_name, author, src):
    zip_filename = download_package(author, package_name)
    with ZipFile(zip_filename, "r") as zip_file:
        ext = zip_file.extract(src)
    dest = SUNPY_EXTERN_DIR / f"{package_name}.py"
    if os.path.exists(dest):
        os.remove(dest)
    shutil.move(ext, dest)


for package_name, (author, src) in PACKAGES.items():
    update_package(package_name, author, src)
```

Reply:
> 
```py
def update_package(package_name, author, src):
    zip_filename = download_package(author, package_name)
    with ZipFile(zip_filename, "r") as zip_file:
        ext = zip_file.extract(src)
    dest = SUNPY_EXTERN_DIR / f"{package_name}.py"
    if os.path.exists(dest):
        os.remove(dest)
    shutil.move(ext, dest)
for package_name, (author, src) in PACKAGES.items():
    update_package(package_name, author, src)
```

Do I have to use this exact or this is just an example?

Reply:
No, it's just an example. It would probably need some other things to make it work properly.

Reply:
Yes, for some reason I thought you want me to use this code.😆

Reply:
I don't see why it wouldn't be simple enough to extract it and then rename it? We can also add that info to the DICT or a named tuple.

Reply:
Because for example, in case of inflect the file name is __init__.py and not inflect.py

Reply:
> Right. But we already have names for them in sunpy.

Yes, doing this will give us the same result, because `__init__.py` will be renamed as `inflect.py` & `distro.py`, `parse.py`, and `appdirs.py` will remain the same.

Reply:
> > Right. But we already have names for them in sunpy.
> 
> Yes, doing this will give us the same result, because `__init__.py` will be renamed as `inflect.py` & `distro.py`, `parse.py`, and `appdirs.py` will remain the same.

Sure, is that a problem?

Reply:
> Sure, is that a problem?

No, I will just do these changes and then commit them.